### DISCUSSION TITLE
Update multi-picker in main screen from modal after row creation in modal

### DISCUSSION THREAD

**Discussion by sdsys-ch** [Budibase/budibase#12153] [Help] [answered]
Hi,

I'm new to Budibase and despite excellent documentation, reading through diverse issues on github (on modals, state etc.) and lots of trying, I'm stuck with what I'm trying to build.

I've got two Postgres tables A and B in the same database. There's a many-to-many relationship between two columns and the relationship is configured.

What I'm trying to do:

- Create a new entry in table A that references one or many rows from Table B in one of it's fields. Often, selecting from the existing options in table B via the multi-picker/many-to-many binding isn't enough and a new entry in table B needs to be created before the new entry to table A can be inserted.

How I'm currently doing this:

The Screen (route: _.../table_a/row/new/_ ) has 

- a form with schema from table A and of type "Create"
- a fieldgroup in which the  multi-picker is set up for entries frmo table B via the defined relationship (existing options show).
- a button inside the fieldgroup ("Create New Entry in Table B") with an action set to navigate to .../table_b/row/new/{{ Form.Fields.entity_name }} . The URL value is for better UI and not used beyond that at the moment. This is opened inside a Modal window.

The Modal (route: _.../table_b/row/new/:entity_name_ ) has

- a form with schema from table B and of type "Create"
- uses `{{ URL.entity_name }}` only for usability / context
- a "Save" button that has the default actions AND an additional "Update State" setting a value to capture the primary key of the newly created row from table B, e.g. `{{ Action 2.Saved row.id }}`. The value is not persisted.
- the Modal is set to simply close (no URL redirect/ ... taking place).

Once the modal closes, the main screen is not updating the multi-picker options. 
**It is however, if I run a search inside the multi-picker (won't find it right away), open the drop-down again and then it's there(!) for the picking. This seems weird/ buggy?**

Despite reading [issue #5629](https://github.com/Budibase/budibase/issues/5629), and more specifically [discussion #5524](https://github.com/Budibase/budibase/discussions/5524) and [pull request #8016](https://github.com/Budibase/budibase/pull/8016) and related, I'm not sure how to actually pull it off in order to either

a) refresh the multi-picker for a possibility to manually choose the newly created option without running a search first
b) add the {{ Action 2.Saved row.id }} from the Modal's state-update to the value(s) already selected in the multi-picker in order to fill table A's relationship-row with all values
c) and best: select the newly created option from the multi-picker alongside any previously selected

before submitting the form. I was unable to make a URI redirect after the modal closes, e.g. using  _.../entity/new_/:tableb_newid_ as default route for the table A Create action, work without loosing the entire form state as filled out until then.

I'm aware that if the entry in table A already existed or if I were to use only screens, this should work with state and URI parameters. This is not an option I believe because creating an entry in table A will usually result in three or more additional tables being updated the same way as above before the actual entry can be inserted to the database.

I appreciate any help or pointers! 

To all devs: kudos and thanks. Budibase is wonderful. Couple hours only and it's already found a spot on my favoured CRUD / automation / time-savers / ... tools list.

-- 
budibase version: 2.11.41
docker engine: 24.0.2
uname: Darwin Kernel Version 22.5.0

**Comment by sdsys-ch** [comment]
Hi,

I've also tested this with the internal database and the result is the same.

I've attached an example app based on the teacher-students example. You can observe above behaviour by going to the first screen (route: _/bb/class/row/new_), input class name, click "New Teacher", Modal opens for route _/bb/teacher/row/new_, input name of teacher (e.g. "Claudia Shannon") and click save, modal closes, open multi-picker "Teacher", "Claudia Shannon" is not in the options list, click the Search bar for the picker, enter "Cla" and wait, it turns up in the options. It is now clickable and can be added to option selection.

Slight difference to the Postgres version: With postgres, the Search will not find the new option unless you search and then close the search bar. After which it can be found in the selectable options via the picker drop-down.

If I can add anything to this or have some basic misconception of how things work, please let me know.

[Test multi picker update-export-1698139210272.tar.gz](https://github.com/Budibase/budibase/files/13113082/Test.multi.picker.update-export-1698139210272.tar.gz)

**Comment by ConorWebb96** [comment]
Hey @sdsys-ch,

I've had an in-depth look at this. Initially I was thinking along the lines of using the `Refresh Data provider` action. I then realised that these are create forms and this wouldn't work. I couldn't find any strange or hack methods to get this to work. 

I do feel your use case is something that should be in the core product. I really liked your idea of passing state back from the modal page by persisting it. So if you can I'd recommend creating a [feature request](https://github.com/Budibase/budibase/issues/new/choose) that would add an extra feature to the screen e.g. `On state change`. (Similar to on load except this listens for state changes). There may be other amends that are better but in my mind this would be a great feature and would possibly allow you to achieve what you're after.

**Reply by sdsys-ch** [reply]
Hi @ConorWebb96 ,

thank you for your time to look into this. I've also played with it a little more but haven't found a good solution.

Having an "On State-change" function would be a good solution for this (and possibly other nice cases). I've opened a feature request and hope this might get enough support soonish. It would make budibase's features stronger and be a godsend for usability cases. Also, I like the idea of having a json data structure in a State variable that keeps track of the user's journey and budibase can react to any changes.

I've [opened an issue](https://github.com/Budibase/budibase/issues/12183). Feel free to comment / ... 

Thanks!

**Reply by sdsys-ch** [reply, marked as answer]
After a comment by @aptkingston in the related [issue](https://github.com/Budibase/budibase/issues/12183), I've had a go again and came up with a hackish way to do it as best as currently possible. See [attached](https://github.com/Budibase/budibase/files/13221275/Multi.picker.update.from.modal.add.row.working-export-1698785118175.tar.gz).

It doesn't do a .selected = true on the newly created option once the modal closes but everything else works as I needed it.

Maybe this helps someone else. Thanks for the help to get there!
